### PR TITLE
Added waggledance metrics with tags for enabling filtering in datadog

### LINKED ISSUE #234: improve metrics by tagging the call and metastore.
Prometheus loves tags, we should tag our metrics with the function call and the metastore. 
We currently put them all in the name this isn't great and hard to create filters in Grafana.

example metric name:
`counter_com_hotels_bdp_waggledance_server_FederatedHMSHandler_get_functions_primary_calls_total
`
h2. Acceptance Criteria:
* counter and duration names are kept with class name e.g: `counter_com_hotels_bdp_waggledance_server_FederatedHMSHandler_calls_total`
* the function call name and metastore name is added as a tag.

### PR DESCRIPTION
**Pull request by DhrubajyotiSadhu**
### :link: Related Issues

fixes #234

### CODE REVIEW

**Review by RickArt — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 1]
```diff
@@ -10,7 +9,7 @@
 
   <groupId>com.hotels</groupId>
   <artifactId>waggle-dance-parent</artifactId>
-  <version>3.11.3-SNAPSHOT</version>
+  <version>3.11.4-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (DhrubajyotiSadhu) [reply pom.xml:12]: Yanked it as there is already a release

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java [hunk 2]
```diff
@@ -51,6 +53,8 @@ public Object monitor(ProceedingJoinPoint pjp) throws Throwable {
   public Object monitor(ProceedingJoinPoint pjp, Monitored monitored) throws Throwable {
 
     String metricBasePath = buildMetricBasePath(pjp);
+    String newMetricBasePath = buildMetricWithTagsBasePath(pjp);
```
[the remark this answers is not in the document]
Comment (DhrubajyotiSadhu) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java]: Done. Thanks for the suggestion.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java [hunk 3]
```diff
@@ -75,21 +85,56 @@ void setMeterRegistry(MeterRegistry meterRegistry) {
...
-    String className = clean(pjp.getSignature().getDeclaringTypeName());
+  private Tags getMetricsTags(String methodName) {
+    Tag federationTag = Tag.of("federation_namespace", getMonitorMetastore());
+    Tag methodTag = Tag.of("method_name", methodName);
+    return Tags.of(federationTag).and(methodTag);
+  }
+
+  private String getMethodName(ProceedingJoinPoint pjp) {
     String methodName = clean(pjp.getSignature().getName());
+    return new StringBuilder(methodName).toString();
```
[the remark this answers is not in the document]
Comment (DhrubajyotiSadhu) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java]: resolved it.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,4 +1,14 @@
-## [3.11.3] - TBD
+## [3.11.4] - 2023-08-16
+### Changed
+- Metrics have been incorporated into Waggle Dance with the inclusion of tags, which will facilitate filtering within Datadog.
+- Exclude jetty-all from core module, because it makes spring start fail and makes `WaggleDanceIntegrationTest` fail.
```
Comment (abhimanyugupta07) [inline CHANGELOG.md:4]: don't think we need to add the previous changelog here.

cc @patduin what you think

File: pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (abhimanyugupta07) [reply pom.xml:12]: The release was made in error.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java [hunk 5]
```diff
@@ -75,21 +85,54 @@ void setMeterRegistry(MeterRegistry meterRegistry) {
...
+
+  private String getMethodName(ProceedingJoinPoint pjp) {
+    return clean(pjp.getSignature().getName());
+  }
+
+  private String getClassName(ProceedingJoinPoint pjp) {
+    return clean(pjp.getSignature().getDeclaringTypeName());
+  }
+
+  private String buildMetricWithTagsBasePath(ProceedingJoinPoint pjp) {
```
Comment (abhimanyugupta07) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java:128]: I don't see this function used actually.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java [hunk 6]
```diff
@@ -67,6 +71,12 @@ public Object monitor(ProceedingJoinPoint pjp, Monitored monitored) throws Throw
       increment(buildMetricPath(COUNTER, metricBasePath, getMonitorMetastore(), result));
```
Comment (abhimanyugupta07) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java:71]: If we are always going to send both metrics (with and without tags), I don't see much reason for having two kind of methods, increment, incrementWithTags, submit, submitWithTags. 

We can just add the functionality to increment and submit methods. You can add the comment "Sends metrics with Tags: federation_namespace and method_name" in there.

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/metrics/MonitoredAspectTest.java [hunk 7]
```diff
@@ -109,6 +115,23 @@ public void monitorSuccesses() throws Throwable {
     assertThat(rs.timer().count(), is(1L));
   }
 
+  @Test
+  public void monitorSuccessWithTags() throws Throwable {
+    aspect.monitor(pjp, monitored);
+
+    Collection<Meter> meters = meterRegistry.get("counter.Type_Anonymous.success").meters();
+    assertThat(meters.size(), is(1));
+    assertThat(((CumulativeCounter) ((ArrayList) meters).get(0)).count(), is(1.0));
```
Comment (patduin) [inline waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/metrics/MonitoredAspectTest.java:124]: this double cast is very hard to read, why don't you follow the examples from the other tests?
```
 RequiredSearch rs = meterRegistry.get("counter.Type_Anonymous.success");
 assertThat(rs.counter().count(), is(1.0));
 ```

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/metrics/MonitoredAspectTest.java [hunk 8]
```diff
@@ -109,6 +115,23 @@ public void monitorSuccesses() throws Throwable {
     assertThat(rs.timer().count(), is(1L));
   }
 
+  @Test
+  public void monitorSuccessWithTags() throws Throwable {
```
Comment (patduin) [inline waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/metrics/MonitoredAspectTest.java:119]: We're not testing the tags, please add asserts that the correct tags are stored.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java [hunk 9]
```diff
@@ -75,21 +85,54 @@ void setMeterRegistry(MeterRegistry meterRegistry) {
...
+    return clean(pjp.getSignature().getName());
+  }
+
+  private String getClassName(ProceedingJoinPoint pjp) {
+    return clean(pjp.getSignature().getDeclaringTypeName());
+  }
+
+  private String buildMetricWithTagsBasePath(ProceedingJoinPoint pjp) {
+    String className = getClassName(pjp);
+    return new StringBuilder(className).toString();
```
[the remark this answers is not in the document]
Comment (patduin) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java:130]: correct, just `return className`

[the remark this answers is not in the document]
Comment (patduin) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/metrics/MonitoredAspect.java:130]: but better to delete the whole method as it is unused

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/metrics/MonitoredAspectTest.java [hunk 10]
```diff
@@ -109,6 +115,23 @@ public void monitorSuccesses() throws Throwable {
     assertThat(rs.timer().count(), is(1L));
```
Comment (patduin) [inline waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/metrics/MonitoredAspectTest.java:115]: Please also add asserts to this integration test: https://github.com/ExpediaGroup/waggle-dance/blob/main/waggle-dance-integration-tests/src/test/java/com/hotels/bdp/waggledance/WaggleDanceIntegrationTest.java#L403

**Comment by massdosage** [comment]
I have a question at a higher level - how many tags will this add to each metric? Bear in mind that Datadog calculates the cost of custom metrics (more or less) as: "metric name X host X tag" so each tag you add is going to increase your custom metric costs. I'm just checking whether this will cause a cost issue or the cardinality of the tag values will be low enough that this won't be an issue.

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,4 +1,14 @@
-## [3.11.3] - TBD
+## [3.11.4] - 2023-08-16
```
Comment (patduin) [inline CHANGELOG.md:1]: update the date please

**Review by patduin — APPROVED** [review]

**Review by RickArt — APPROVED** [review]

**Pull request merged**